src/nodes/ipv6_nd_node.c

```diff
@@ -55,37 +52,39 @@ static __rte_always_inline rte_edge_t get_next_index(__rte_unused struct rte_nod
 	struct rte_ipv6_hdr *req_ipv6_hdr = (struct rte_ipv6_hdr *)(req_eth_hdr + 1);
 	struct icmp6hdr *req_icmp6_hdr = (struct icmp6hdr *)(req_ipv6_hdr + 1);
 	uint8_t icmp_type = req_icmp6_hdr->icmp6_type;
-	const uint8_t *rt_ip = dp_get_gw_ip6();
+	const union dp_ipv6 *gw_ip = dp_get_gw_ip6();
 	struct nd_msg *nd_msg;
 	struct ra_msg *req_ra_msg;
 	struct dp_port *port = dp_get_in_port(m);
+	const union dp_ipv6 *src_ipv6 = dp_get_src_ipv6(req_ipv6_hdr);
 
 	rte_ether_addr_copy(&req_eth_hdr->src_addr, &req_eth_hdr->dst_addr);
 	rte_ether_addr_copy(&port->own_mac, &req_eth_hdr->src_addr);
 
-	if (!memcmp(req_ipv6_hdr->src_addr, dp_unspecified_ipv6, sizeof(req_ipv6_hdr->src_addr)))
-		rte_memcpy(req_ipv6_hdr->dst_addr, dp_multicast_ipv6, sizeof(req_ipv6_hdr->dst_addr));
+	if (dp_ipv6_match(src_ipv6, &dp_empty_ipv6))
+		dp_set_dst_ipv6(req_ipv6_hdr, &dp_multicast_ipv6);
 	else
-		rte_memcpy(req_ipv6_hdr->dst_addr, req_ipv6_hdr->src_addr, sizeof(req_ipv6_hdr->dst_addr));
+		dp_set_dst_ipv6(req_ipv6_hdr, src_ipv6);
 
-	rte_memcpy(req_ipv6_hdr->src_addr, rt_ip, sizeof(req_ipv6_hdr->src_addr));
+	dp_set_src_ipv6(req_ipv6_hdr, gw_ip);
 
 	if (icmp_type != NDISC_NEIGHBOUR_SOLICITATION && icmp_type != NDISC_ROUTER_SOLICITATION)
 		return IPV6_ND_NEXT_DROP;
 
 	if (icmp_type == NDISC_NEIGHBOUR_SOLICITATION) {
 		nd_msg = (struct nd_msg *)(req_ipv6_hdr + 1);
-		if (memcmp(&nd_msg->target, rt_ip, sizeof(nd_msg->target)))
+		static_assert(sizeof(nd_msg->target) == sizeof(gw_ip->bytes), "Incompatible IPv6 format in ND message structure");
```

Comment:
Exactly, it's just a paranoia check (or maybe Rust-like check?) since the `memcmp` comparison is "unchecked" in nature, so at least something is there and should be a compile-time thing, so why not just add it.